// @postman/postman-mcp-server/dist/src/tools/getSpecDefinition.js
import { z } from 'zod';
import { asMcpError, McpError } from './utils/toolHelpers.js';
export const method = 'getSpecDefinition';
export const description = "Gets the complete contents of an OpenAPI or AsyncAPI specification's definition.";
export const parameters = z.object({ specId: z.string().describe("The spec's ID.") });
export const annotations = {
    title: "Gets the complete contents of an OpenAPI or AsyncAPI specification's definition.",
    readOnlyHint: true,
    destructiveHint: false,
    idempotentHint: true,
};
export async function handler(args, extra) {
    try {
        const endpoint = `/specs/${args.specId}/definitions`;
        const query = new URLSearchParams();
        const url = query.toString() ? `${endpoint}?${query.toString()}` : endpoint;
        const options = {
            headers: extra.headers,
        };
        const result = await extra.client.get(url, options);
        return {
            content: [
                {
                    type: 'text',
                    text: `${typeof result === 'string' ? result : JSON.stringify(result, null, 2)}`,
                },
            ],
        };
    }
    catch (e) {
        if (e instanceof McpError) {
            throw e;
        }
        throw asMcpError(e);
    }
}
